keras_rcnn/preprocessing/_object_detection.py

```diff
@@ -5,80 +5,83 @@
-        metadata       = numpy.zeros((batch_size, 3),                   dtype=keras.backend.floatx())
+        # Labels has num_classes + 1 elements, since 0 is reserved for background.
+        num_classes = len(self.classes.keys())
+        images      = numpy.zeros((batch_size,) + self.target_shape, dtype=keras.backend.floatx())
+        boxes       = numpy.zeros((batch_size, 0, 4),                dtype=keras.backend.floatx())
+        labels      = numpy.zeros((batch_size, 0, num_classes + 1),  dtype=numpy.uint8)
 
         for batch_index, image_index in enumerate(selection):
-            path         = self.data[image_index]["filename"]
-            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
-            image, scale = scale_image(image)
-            image        = self.image_data_generator.random_transform(image)
-            image        = self.image_data_generator.standardize(image)
-
-            # Copy image to batch blob
-            image_batch[batch_index] = image
-
-            # Set ground truth boxes
-            boxes = self.data[image_index]["boxes"]
-            for i, b in enumerate(boxes):
+            path  = self.dictionary[image_index]["filename"]
+            image = skimage.io.imread(path)
+
+            # Assert that the loaded image has the predefined image shape.
+            if image.shape != self.image_shape:
+                raise Exception("All input images need to be of the same shape.")
+
+            # Copy image to batch blob.
+            images[batch_index] = skimage.transform.rescale(image, scale=self.scale, mode="reflect")
+
+            # Set ground truth boxes.
+            for i, b in enumerate(self.dictionary[image_index]["boxes"]):
                 if b["class"] not in self.classes:
                     raise Exception("Class {} not found in '{}'.".format(b["class"], self.classes))
 
-                gt_data = [b["y1"], b["x1"], b["y2"], b["x2"], self.classes[b["class"]]]
-                gt_boxes_batch = numpy.append(gt_boxes_batch, [[gt_data]], axis=1)
+                box   = [b["y1"], b["x1"], b["y2"], b["x2"]]
+                boxes = numpy.append(boxes, [[box]], axis=1)
 
-            # Scale the ground truth boxes to the selected image scale
-            gt_boxes_batch[batch_index, :, :4] *= scale
+                # Store the labels in one-hot form.
+                label = [0] * (num_classes + 1)
+                label[self.classes[b["class"]]] = 1
+                labels = numpy.append(labels, [[label]], axis = 1)
 
-            # Create metadata
-            metadata[batch_index, :] = [image.shape[0], image.shape[1], scale]
+            # Scale the ground truth boxes to the selected image scale.
+            boxes[batch_index, :, :4] *= self.scale
 
-        return [image_batch, gt_boxes_batch, metadata], None
+        return [images, boxes, labels, self.scale], None
```

Comment:
Generators must return `X, Y`, but `Y` doesn't make sense in our case because the network generates `Y`. Therefore, `Y` can be ignored.